### PR TITLE
Toolbar Dragging

### LINKED ISSUE #1494: Bid history not centered underneath the bid button
image

### PR DESCRIPTION
**Pull request by orta**
The only commit is this : https://github.com/artsy/eigen/commit/312403c3d6e4f2d2c41688c1d5de0249514c15e7

This adds the initial implementation of Toolbar dragging for native live auctions, this feature ends up creating some tricky design constraints when mixed with the ability to page sideways. So, I'm leaving this here as a stub, I will come back to it once I've taken a second shot at the improved image previews.

Fixes #1494.
Related to #1118.

### CODE REVIEW

**Comment by ashfurrow** [comment]
Approach makes sense 👍

**Comment by ArtsyMobile** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-kind="Warning">
        
            3 Warnings
        
      </th>
     </tr>
  </thead>
  <tbody>
        <tr>
      <td>:warning:</td>
      <td data-sticky="true">PR is classed as Work in Progress</td>
    </tr>
        <tr>
      <td>:warning:</td>
      <td data-sticky="true">Needs testing on a Phone if change is non-trivial</td>
    </tr>
        <tr>
      <td>:warning:</td>
      <td data-sticky="true">Detected some Swift building time outliers</td>
    </tr>
    
    
  </tbody>
</table>
  

  



Time | Class | Function |
| --- | ----- | ----- |253.8ms | [AuctionViewController.swift](/artsy/eigen/tree/orta-toolbar_dragging/Artsy/View_Controllers/Auction/AuctionViewController.swift#L176) | @objc dynamic func showRefineTappedAnimated(animated: Bool)
194.6ms | [SwiftyJSON.swift](/artsy/eigen/tree/orta-toolbar_dragging/Pods/SwiftyJSON/Source/SwiftyJSON.swift#L445) | get {}
163.5ms | [WebSocket.swift](/artsy/eigen/tree/orta-toolbar_dragging/Pods/Starscream/Source/WebSocket.swift#L528) | private func processRawMessage(buffer: UnsafePointer<UInt8>, bufferLen: Int)




<p align="right" data-meta="generated_by_danger" data-base-commit="" data-head-commit="" >
  Generated by :no_entry_sign: <a href="https://github.com/danger/danger/">danger</a>
</p>

**Comment by ashfurrow** [comment]
Taking over this branch, I've merged from master. Will update when I get work completed.

**Comment by ashfurrow** [comment]
Okaaaaay, so this is ready for review. 




This is shippable, but has a few rough edges. Mainly: 

- What happens when a user is looking at the bid history, and the current lot changes? @katarinabatina ?
- The user is currently able to "scroll up" to see bid history that isn't there.

These issues' code is related to #1538 and I'll probably wrap them up together. For now, I want to get this reviewed and merged so I don't diverge for too long.

File: Artsy/View_Controllers/Live_Auctions/LiveAuctionViewController.swift [hunk 1]
```diff
@@ -156,7 +156,7 @@ extension PrivateFunctions {
     }
 
     func salesPerson(sale: LiveSale, jwt: JWT, bidderID: String?) -> LiveAuctionsSalesPersonType {
-        return LiveAuctionsSalesPerson(sale: sale, jwt: jwt, bidderID: bidderID)
+        return LiveAuctionsSalesPerson(sale: sale, jwt: jwt, bidderID: bidderID, stateManagerCreator: LiveAuctionsSalesPerson.stubbedStateManagerCreator())
```
Comment (ashfurrow) [inline Artsy/View_Controllers/Live_Auctions/LiveAuctionViewController.swift]: Ha, probably want to remove this before merging :wink: Going to wait for other feedback and roll that change in.

File: Artsy/View_Controllers/Live_Auctions/LiveAuctionLotViewController.swift [hunk 2]
```diff
@@ -176,9 +178,124 @@ class LiveAuctionLotViewController: UIViewController {
...
+        }
+    }
+
+    // This is strictly iPad support, trait collections on iPhone won't chage as we don't support rotation.
+    override func viewWillTransitionToSize(size: CGSize, withTransitionCoordinator coordinator: UIViewControllerTransitionCoordinator) {
+        super.viewWillTransitionToSize(size, withTransitionCoordinator: coordinator)
+
+        coordinator.animateAlongsideTransition({ context in
+
+            self.alignMetadataToTopConstraint?.constant = self.openedMetadataPosition ?? 0 // Reset this to stick to the top, we'll set its active status below.
```
Comment (orta) [inline Artsy/View_Controllers/Live_Auctions/LiveAuctionLotViewController.swift]: this chunk of code isn't immediately obvious what is happening, maybe worth saying something like "on rotation, close the bid history"

File: Artsy/View_Controllers/Live_Auctions/LiveAuctionLotViewController.swift [hunk 3]
```diff
@@ -176,9 +178,124 @@ class LiveAuctionLotViewController: UIViewController {
...
+    private var _bidHistoryState: BidHistoryState = .Closed {
+        willSet(newValue) {
+            if newValue != _bidHistoryState {
+                self.bidHistoryState.update(newValue)
+            }
+        }
+    }
+    private var _initialBidHistoryState: BidHistoryState = .Closed
+
+    func dragToolbar(gesture: UIPanGestureRecognizer) {
```
Comment (orta) [inline Artsy/View_Controllers/Live_Auctions/LiveAuctionLotViewController.swift]: `dragToolbar` do you feel this name still describes what's going on?

File: Artsy/View_Controllers/Live_Auctions/LiveAuctionViewController.swift (hunk 1 above)
Comment (orta) [inline Artsy/View_Controllers/Live_Auctions/LiveAuctionViewController.swift]: hah

File: Artsy/View_Controllers/Live_Auctions/LotListCollectionViewCell.swift [hunk 4]
```diff
@@ -20,8 +20,13 @@ class LotListCollectionViewCell: UICollectionViewCell {
...
     private var userInterfaceNeedsSetup = true
-    private var lotStateSubscription: (ObserverToken, Observable<LotState>)?
-    private var askingPriceSubscription: (ObserverToken, Observable<UInt64>)?
+    private var lotStateSubscription: ObserverToken<LotState>?
+    private var askingPriceSubscription: ObserverToken<UInt64>?
+
+    deinit {
+        lotStateSubscription?.unsubscribe()
+        askingPriceSubscription?.unsubscribe()
+    }
```
Comment (orta) [inline Artsy/View_Controllers/Live_Auctions/LotListCollectionViewCell.swift]: nice

File: Artsy/View_Controllers/Live_Auctions/Views/LiveAuctionFancyLotCollectionViewLayout.swift [hunk 5]
```diff
@@ -132,20 +137,23 @@ private extension PrivateFunctions {
...
         case 2 where ratioDragged < -0.5:
-            applyFancyLayoutToAttributes(&layoutAttributes, position: .PreviousUnderflow)
+            applyFancyLayoutToAttributes(&copy, position: .PreviousUnderflow)
         case 2:
-            applyFancyLayoutToAttributes(&layoutAttributes, position: .Next)
+            applyFancyLayoutToAttributes(&copy, position: .Next)
         default: break;
         }
+
+        return copy
```
Comment (orta) [inline Artsy/View_Controllers/Live_Auctions/Views/LiveAuctionFancyLotCollectionViewLayout.swift]: nice

File: Artsy/View_Controllers/Live_Auctions/Views/LiveAuctionFancyLotCollectionViewLayout.swift [hunk 6]
```diff
@@ -71,9 +78,7 @@ class LiveAuctionFancyLotCollectionViewLayout: UICollectionViewFlowLayout {
     }
 
     override func layoutAttributesForItemAtIndexPath(indexPath: NSIndexPath) -> UICollectionViewLayoutAttributes? {
-        return super.layoutAttributesForItemAtIndexPath(indexPath)?.then { layoutAttributes in
-            modifyLayoutAttributes(&layoutAttributes)
-        }
+        return super.layoutAttributesForItemAtIndexPath(indexPath).flatMap { modifiedLayoutAttributesCopy($0) }
```
Comment (orta) [inline Artsy/View_Controllers/Live_Auctions/Views/LiveAuctionFancyLotCollectionViewLayout.swift]: elegant

File: Artsy/View_Controllers/Live_Auctions/Views/LiveAuctionFancyLotCollectionViewLayout.swift [hunk 7]
```diff
@@ -275,4 +287,36 @@ private extension PrivateFunctions {
...
+    func applyRepulsionToMetrics(metrics: LayoutMetrics, atPosition position: CellPosition, aspectRatio: CGFloat) -> LayoutMetrics {
+        switch position {
+        case .Current where aspectRatio > (maxCurrentWidth / maxCurrentHeight):
+            // Modify height
+            return (restingWidth: metrics.restingWidth, restingHeight: metrics.restingHeight - (repulsionConstant / aspectRatio), targetWidth: metrics.targetWidth, targetHeight: metrics.targetHeight - (repulsionConstant / aspectRatio))
+        case .Current: // aspectRatio <= (maxCurrentWidth / maxCurrentHeight)
+            // Modify width
+            return (restingWidth: metrics.restingWidth - (repulsionConstant * aspectRatio), restingHeight: metrics.restingHeight, targetWidth: metrics.targetWidth - (repulsionConstant * aspectRatio), targetHeight: metrics.targetHeight)
+        default: return metrics
+        }
```
Comment (orta) [inline Artsy/View_Controllers/Live_Auctions/Views/LiveAuctionFancyLotCollectionViewLayout.swift]: can't _quite_ decide if the `where` clause makes it harder to understand of not here

File: Artsy/View_Controllers/Live_Auctions/LiveAuctionLotViewController.swift (hunk 3 above)
Comment (ashfurrow) [inline Artsy/View_Controllers/Live_Auctions/LiveAuctionLotViewController.swift]: nah, I'll change it

**Comment by orta** [comment]
All in all - feels _great_ - I've seen this in person and it feels awesome.

**Comment by ashfurrow** [comment]
Thanks! I'll send up the feedback commits momentarily.

File: Artsy/View_Controllers/Live_Auctions/Views/LiveAuctionFancyLotCollectionViewLayout.swift (hunk 7 above)
Comment (ashfurrow) [inline Artsy/View_Controllers/Live_Auctions/Views/LiveAuctionFancyLotCollectionViewLayout.swift]: yeah, hard to tell. I've seen it a lot in causality, seems to be a more common use there.

**Comment by ashfurrow** [comment]
It's green! 🍏 @orta care to merge?

**Comment by orta** [comment]
I would

**Comment by ashfurrow** [comment]
You've made my day :bow:

**Pull request merged**